src/filter/json_encode_filter.rs

```diff
@@ -0,0 +1,268 @@
+            })
+        );
+    }
+
+    #[test]
+    fn parsable_line_parsing_on() {
+        // Test we do parse a line if parsing is on
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "{\"good\": \"do parse\"}",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "good": "do parse",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    #[test]
+    fn unparsable_line() {
+        // Test we don't parse a line if it's not JSON
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "this is not json",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "value": "this is not json",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    #[test]
+    fn non_object_line() {
+        // Test we don't parse a line if it's not a JSON object but is valid JSON
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "[123, \"not an object\"]",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "value": "[123, \"not an object\"]",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    // quickcheck and serde_json::map::Map aren't compatible, so we ask quickcheck for
+    // many Vec<(String, String)>s that we turn into maps.
+    fn vecs_to_objs(vecs: &Vec<Vec<(String, String)>>) -> Vec<Map<String, Value>> {
+        iter(vecs).map(|vec|
+            iter(vec).map(|(k, v)| (k.clone(), v.clone().into()))
+        ).into()
+    }
+
+    quickcheck! {
+        fn merged_objects_contain_all_keys(vecs: Vec<Vec<(String, String)>>) -> bool {
+            let result = merge_objects(&vecs_to_objs(&vecs));
+            for obj in vecs {
+                for (k, v) in obj {
+                    if !result.contains_key(k) {
+                        return false
+                    }
+                }
+            }
+            true
+        }
+
+        fn merged_objects_takes_first_value(vecs: Vec<Vec<(String, String)>>) -> bool {
+            let objs = vecs_to_objs(&vecs);
+            let result = merge_objects(&objs);
+            for (key, result_value) in result {
+                match iter(objs).find(|obj| obj.contains_key(key)) {
```

Comment:
This `iter(objs)` is going to cause a build error I do believe.

Reply:
yep, there's a bunch of build issues i'm just dealing with now, i was trying to parallelise fixing those with higher-level feedback :P